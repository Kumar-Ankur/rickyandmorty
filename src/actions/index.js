import * as actionTypes from "../types";

/**
 * @function getCharacters
 * @description action types for fetching characters data from API using saga file and update the state.
 */
export function getCharacters() {
  return { type: actionTypes.GET_CHARACTERS };
}

/**
 * @function getFilteredData
 * @param { Object } data - Filtered data
 * @description - filtered data based on payload value.
 */
export function getFilteredData(data) {
  return { type: actionTypes.GET_FILTERED_DATA, payload: data };
}

/**
 * @function sortAscendingById
 * @description - sort all the character data in ascending order
 */
export function sortAscendingById() {
  return { type: actionTypes.SORT_DATA_BY_ASCENDING };
}

/**
 * @function sortDescendingById
 * @description - sort all the character data in descending order.
 */
export function sortDescendingById() {
  return { type: actionTypes.SORT_DATA_BY_DESCENDING };
}

/**
 * @function addFilteredData
 * @param { String } data
 * @description - add filtered data in string for species category state
 */
export function addFilteredData(data) {
  return { type: actionTypes.ADD_FILTERED_DATA, payload: data };
}

/**
 * @function removeFilteredData
 * @param { String } data
 * @description - remove filtered data in string for species category state
 */
export function removeFilteredData(data) {
  return { type: actionTypes.REMOVE_FILTERED_DATA, payload: data };
}

/**
 * @function addGenderFilteredData
 * @param { String } data
 * @description - add filtered data in string for gender category state
 */
export function addGenderFilteredData(data) {
  return { type: actionTypes.ADD_GENDER_FILTERED_DATA, payload: data };
}

/**
 * @function removeGenderFilteredData
 * @param { String } data
 * @description - remove filtered data in string for gender category state
 */
export function removeGenderFilteredData(data) {
  return { type: actionTypes.REMOVE_GENDER_FILTERED_DATA, payload: data };
}

/**
 * @function addOriginFilteredData
 * @param { String } data
 * @description - add filtered data in string for origin category state
 */
export function addOriginFilteredData(data) {
  return { type: actionTypes.ADD_ORIGIN_FILTERED_DATA, payload: data };
}

/**
 * @function removeOriginFilteredData
 * @param { String } data
 * @description - remove filtered data in string for origin category state
 */
export function removeOriginFilteredData(data) {
  return { type: actionTypes.REMOVE_ORIGIN_FILTERED_DATA, payload: data };
}

/**
 * @function addSearchText
 * @param { String } data
 * @description - add search data in string for search category state
 */
export function addSearchText(data) {
  return { type: actionTypes.ADD_SEARCH_TEXT, payload: data };
}
