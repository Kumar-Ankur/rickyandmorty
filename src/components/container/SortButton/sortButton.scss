.btn {
  padding: 0.5rem 0.5rem;
  text-transform: uppercase;
  text-decoration: none;
  display: inline-block;
  border-radius: 10rem;
  transition: all 0.2s;
  position: relative;
  cursor: pointer;

  &:hover {
    transform: translateY(-0.3rem);
    box-shadow: 0 10px 20px rgba(black, 0.2);

    &::after {
      transform: scaleX(1.4) scaleY(1.6);
      opacity: 0;
    }
  }

  &--white {
    background-color: white;
    color: black;

    &::after {
      background-color: white;
    }
  }

  &::after {
    content: "";
    display: inline-block;
    height: 100%;
    width: 100%;
    border-radius: 10rem;
    position: absolute;
    top: 0;
    left: 0;
    z-index: -1;
    transition: all 0.4s;
  }

  &--animated {
    animation: moveInBottom 1s ease-out 0.75s;
    animation-fill-mode: backwards;
  }
}

@keyframes moveInBottom {
  0% {
    opacity: 0;
    transform: translateY(3rem);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}
