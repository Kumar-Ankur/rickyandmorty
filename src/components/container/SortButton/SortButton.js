import React, { Component } from "react";
import { connect } from "react-redux";
import * as actionCreator from "../../../actions";
import * as constants from "../../../constants";
import "./sortButton.scss";

class SortButton extends Component {
  render() {
    return (
      <>
        <button
          data-test="component-btn-asc"
          onClick={() => this.props.sortAscendingById()} // dispatch an action to sort data based on id in ascending order
          type="submit"
          className="btn btn--animated btn--white"
          style={{ marginRight: "1rem" }}
        >
          {constants.SORT_ASCENDING_TEXT}
        </button>
        <button
          data-test="component-btn-desc"
          onClick={() => this.props.sortDescendingById()} // dispatch an action to sort data based on id in descending order
          type="submit"
          className="btn btn--animated btn--white"
        >
          {constants.SORT_DESCENDING_TEXT}
        </button>
      </>
    );
  }
}

export const mapDispatchToProps = (dispatch) => {
  return {
    sortAscendingById: () => dispatch(actionCreator.sortAscendingById()),
    sortDescendingById: () => dispatch(actionCreator.sortDescendingById()),
  };
};

export default connect(null, mapDispatchToProps)(SortButton);
