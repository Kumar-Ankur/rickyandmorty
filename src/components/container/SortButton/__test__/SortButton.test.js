import React from "react";
import { shallow } from "enzyme";
import sinon from "sinon";
import SortButton, { mapDispatchToProps } from "../SortButton";
import { applyMiddleware, createStore } from "redux";
import createSagaMiddleware from "redux-saga";
import rootReducer from "../../../../reducers";
import * as actionCreator from "../../../../actions";

const middleware = createSagaMiddleware();
const mockStore = createStore(rootReducer, applyMiddleware(middleware));

/**
 * Factory function to create a shallowwrapper for the App component.
 * @function setUp
 * @param { Object } props - Component props specific to this setup
 * @param { Object } state - Initial state for setUp
 * @returns { shallowWrapper }
 *
 */
function setUp(props = {}) {
  return shallow(<SortButton store={mockStore} {...props} />);
}

/**
 * Return shallowWarpper containing node(s) with the given data-test attribute
 * @function findByTestAttr
 * @param {ShallowWrapper} wrapper - Enzyme shallow wrapper to search within
 * @param {string} val - value of data-test attribute for search
 * @returns { shallowWrapper}
 */
function findByTestData(wrapper, val) {
  return wrapper.find(`[data-test='${val}']`);
}

describe("check sort button component", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = setUp().dive();
  });

  test("genrate snapshot for sort button component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("check sort ascending button render without error", () => {
    const sortAscBtn = findByTestData(wrapper, "component-btn-asc");
    expect(sortAscBtn.length).toBe(1);
  });

  test("check sort descending button render without error", () => {
    const sortDescBtn = findByTestData(wrapper, "component-btn-desc");
    expect(sortDescBtn.length).toBe(1);
  });

  test("check `sortAscendingById` function call from mapDispatchToProps", () => {
    const dispatchSpy = sinon.spy();
    mockStore.dispatch = dispatchSpy;
    const { sortAscendingById } = mapDispatchToProps(dispatchSpy);
    sortAscendingById();
    const lastSpyCall = dispatchSpy.args[0][0];
    const expectedAction = actionCreator.sortAscendingById();
    expect(lastSpyCall.types).toEqual(expectedAction.types);
  });

  test("check `sortDescendingById` function call from mapDispatchToProps", () => {
    const dispatchSpy = sinon.spy();
    mockStore.dispatch = dispatchSpy;
    const { sortDescendingById } = mapDispatchToProps(dispatchSpy);
    sortDescendingById();
    const lastSpyCall = dispatchSpy.args[0][0];
    const expectedAction = actionCreator.sortDescendingById();
    expect(lastSpyCall.types).toEqual(expectedAction.types);
  });

  test('check onClick method on sortAscendingButton', () => {
    const sortAscBtn = findByTestData(wrapper, "component-btn-asc");
    const sortAscendingByIdMock = jest.fn();
    const instance = wrapper.instance();
    instance.sortAscendingById = sortAscendingByIdMock;
    sortAscBtn.simulate('click');
    instance.sortAscendingById();
    expect(sortAscendingByIdMock).toHaveBeenCalled();
  });

  test('check onClick method on sortAscendingButton', () => {
    const sortAscBtn = findByTestData(wrapper, "component-btn-desc");
    const sortDescendingByIdMock = jest.fn();
    const instance = wrapper.instance();
    instance.sortDescendingById = sortDescendingByIdMock;
    sortAscBtn.simulate('click');
    instance.sortDescendingById();
    expect(sortDescendingByIdMock).toHaveBeenCalled();
  });
});
