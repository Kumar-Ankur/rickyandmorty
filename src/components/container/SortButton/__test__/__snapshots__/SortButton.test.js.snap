// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check sort button component genrate snapshot for sort button component 1`] = `ShallowWrapper {}`;
